File: JS/easy-solution/110.js
/**
110. Balanced Binary Tree
    Given a binary tree, determine if it is height-balanced.

    Example :
    Input: root = [3,9,20,null,null,15,7]
    Output: true
 */


/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function(root) {
    let balanced = true;
    function dfs(node){
        if (node === null)
            return 0;
        const l = dfs(node.left);
        const r = dfs(node.right);
        if (Math.abs(l - r) > 1)
            balanced = false;
        return Math.max(l, r) + 1;
    }
    dfs(root);
    return balanced;
};

File: cpp/easy-solution/110.cc
/**
110. Balanced Binary Tree
    Given a binary tree, determine if it is height-balanced.

    Example :
    Input: root = [3,9,20,null,null,15,7]
    Output: true
 */
#include <iostream>
#include <algorithm>
using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool balanced;
    int dfs(TreeNode* node){
        if (!node)
            return 0;
        int l = dfs(node->left);
        int r = dfs(node->right);
        if (abs(l - r) > 1)
            balanced = false;
        return max(l, r) + 1;
    }
    bool isBalanced(TreeNode* root) {
        balanced = true;
        dfs(root);
        return balanced;
    }
};